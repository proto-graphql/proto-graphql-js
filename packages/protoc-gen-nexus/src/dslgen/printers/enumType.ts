import type { Registry } from "@bufbuild/protobuf";
import {
  type EnumType,
  compact,
  protobufGraphQLExtensions,
} from "@proto-graphql/codegen-core";
import { type Code, code, literalOf } from "ts-poet";

import { impNexus, nexusTypeDef } from "./util.js";

/**
 * @example
 * ```ts
 * export cosnt Hello = enumType({
 *   name: "Hello",
 *   // ...
 * })
 * ```
 */
export function createEnumTypeCode(type: EnumType, registry: Registry): Code {
  const typeOpts = {
    name: type.typeName,
    description: type.description,
    members: type.values.map((ev) => ({
      name: ev.name,
      value: ev.number,
      description: ev.description,
      deprecation: ev.deprecationReason,
      extensions: protobufGraphQLExtensions(ev, registry),
    })),
    extensions: protobufGraphQLExtensions(type, registry),
  };
  return code`export const ${nexusTypeDef(type)} = ${impNexus(
    "enumType",
  )}(${literalOf(compact(typeOpts))});`;
}
