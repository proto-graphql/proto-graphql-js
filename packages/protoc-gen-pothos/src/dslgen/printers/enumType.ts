import type { Registry } from "@bufbuild/protobuf";
import {
  type EnumType,
  compact,
  protoType,
  protobufGraphQLExtensions,
} from "@proto-graphql/codegen-core";
import { type Code, code, imp, joinCode, literalOf } from "ts-poet";

import { type PothosPrinterOptions, pothosBuilder, pothosRef } from "./util.js";

/**
 * @example
 * ```ts
 * export cosnt Hello$Ref: EnumRef<Hello, Hello> = builder.enumType("Hello", {
 *   values: [
 *     // ...
 *   ],
 *   // ...
 * })
 * ```
 */
export function createEnumTypeCode(
  type: EnumType,
  registry: Registry,
  opts: PothosPrinterOptions,
): Code {
  const typeOpts = {
    description: type.description,
    values: code`{${joinCode(
      type.values
        .filter((v) => !v.isIgnored() && !v.isUnespecified())
        .map(
          (ev) =>
            code`${ev.name}: ${literalOf(
              compact({
                description: ev.description,
                deprecationReason: ev.deprecationReason,
                value: ev.number,
                extensions: protobufGraphQLExtensions(ev, registry),
              }),
            )},`,
        ),
    )}} as const`,
    extensions: protobufGraphQLExtensions(type, registry),
  };

  const protoTypeExpr = protoType(type.proto, opts);
  // EnumRef<Hello, Hello>
  const refTypeExpr = code`${imp(
    "EnumRef@@pothos/core",
  )}<${protoTypeExpr}, ${protoTypeExpr}>`;

  return code`
    export const ${pothosRef(type)}: ${refTypeExpr} =
      ${pothosBuilder(type, opts)}.enumType(${literalOf(
        type.typeName,
      )}, ${literalOf(compact(typeOpts))});
  `;
}
