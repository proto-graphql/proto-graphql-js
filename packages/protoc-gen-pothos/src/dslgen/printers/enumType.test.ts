import { createEcmaScriptPlugin } from "@bufbuild/protoplugin";
import {
  EnumType,
  type TypeOptions,
  createRegistryFromSchema,
  defaultScalarMapping,
  defaultScalarMappingForTsProto,
} from "@proto-graphql/codegen-core";
import {
  type TestapisPackage,
  buildCodeGeneratorRequest,
} from "@proto-graphql/testapis-proto";
import { describe, expect, test } from "vitest";
import { printEnumType } from "./enumType.js";
import type { PothosPrinterOptions } from "./util.js";

type TestCase = {
  test: string;
  args: {
    packageName: TestapisPackage;
    enumTypeNameInProto: string;
  };
};

type TestSuite = {
  suite: string;
  options: PothosPrinterOptions;
  cases: TestCase[];
};

const testSuites: TestSuite[] = [
  {
    suite: "ts-proto",
    options: {
      dsl: "pothos",
      protobuf: "ts-proto" as const,
      importPrefix: "@testapis/ts-proto",
      emitImportedFiles: false,
      fileLayout: "proto_file",
      filenameSuffix: ".pothos",
      pothos: {
        builderPath: "../../builder",
      },
    },
    cases: [
      {
        test: "generates code for a simple enum",
        args: {
          packageName: "testapis.enums",
          enumTypeNameInProto: "MyEnum",
        },
      },
      {
        test: "generates code for an enum without unspecified",
        args: {
          packageName: "testapis.enums",
          enumTypeNameInProto: "MyEnumWithoutUnspecified",
        },
      },
      {
        test: "generates code for nested enum",
        args: {
          packageName: "testapis.nested",
          enumTypeNameInProto: "ParentMessage.NestedEnum",
        },
      },
      {
        test: "generates code for enum with extensions",
        args: {
          packageName: "testapis.extensions",
          enumTypeNameInProto: "PrefixedEnum",
        },
      },
    ],
  },
  {
    suite: "protobuf-es",
    options: {
      dsl: "pothos",
      protobuf: "protobuf-es" as const,
      importPrefix: "@testapis/protobuf-es",
      emitImportedFiles: false,
      fileLayout: "proto_file",
      filenameSuffix: ".pothos",
      pothos: {
        builderPath: "../../builder",
      },
    },
    cases: [
      {
        test: "generates code for a simple enum",
        args: {
          packageName: "testapis.enums",
          enumTypeNameInProto: "MyEnum",
        },
      },
      {
        test: "generates code for an enum without unspecified",
        args: {
          packageName: "testapis.enums",
          enumTypeNameInProto: "MyEnumWithoutUnspecified",
        },
      },
      {
        test: "generates code for nested enum",
        args: {
          packageName: "testapis.nested",
          enumTypeNameInProto: "ParentMessage.NestedEnum",
        },
      },
    ],
  },
];

// New function to generate code using protoplugin
function generateEnumTypeWithPrintFunction(
  packageName: TestapisPackage,
  enumTypeNameInProto: string,
  options: PothosPrinterOptions,
): string {
  const typeOptions: TypeOptions = {
    partialInputs: false,
    scalarMapping:
      options.protobuf === "ts-proto"
        ? defaultScalarMappingForTsProto
        : defaultScalarMapping,
    ignoreNonMessageOneofFields: false,
  };

  const plugin = createEcmaScriptPlugin({
    name: "test",
    version: "0.0.0",
    generateTs: (schema) => {
      const registry = createRegistryFromSchema(schema);

      // The actual proto package might differ from the TestapisPackage key
      // For example: "testapis.enums" key but "testapi.enums" proto package
      let descEnum = registry.getEnum(`${packageName}.${enumTypeNameInProto}`);

      if (descEnum === undefined && packageName === "testapis.enums") {
        // Try with the actual proto package name
        descEnum = registry.getEnum(`testapi.enums.${enumTypeNameInProto}`);
      }

      if (descEnum === undefined) {
        throw new Error(
          `Enum ${enumTypeNameInProto} not found in package ${packageName}`,
        );
      }

      const enumType = new EnumType(descEnum, typeOptions);

      const f = schema.generateFile("generated.ts");
      printEnumType(f, enumType, registry, options);
    },
  });

  const req = buildCodeGeneratorRequest(packageName);
  req.parameter = "target=ts";

  const resp = plugin.run(req);

  const file = resp.file.find((f) => f.name === "generated.ts");
  if (!file) {
    throw new Error("Generated file not found");
  }

  return file.content;
}

describe("createEnumTypeCode", () => {
  //   for (const { suite, options, cases } of testSuites) {
  //     describe(suite, () => {
  //       test.each(cases)("$test", ({ args }) => {
  //         const code = generateEnumTypeCode(
  //           args.packageName,
  //           args.enumTypeNameInProto,
  //           options,
  //         );
  //         expect(code).toMatchSnapshot();
  //       });
  //     });
  //   }
  // });
  //
  // describe("printEnumType", () => {
  //   const shouldUsePrintFunction = process.env.USE_PROTOPLUGIN_PRINTER === "1";
  //
  //   if (!shouldUsePrintFunction) {
  //     test.skip("printEnumType tests skipped", () => {});
  //     return;
  //   }

  for (const { suite, options, cases } of testSuites) {
    describe(suite, () => {
      test.each(cases)("$test", ({ args }) => {
        const code = generateEnumTypeWithPrintFunction(
          args.packageName,
          args.enumTypeNameInProto,
          options,
        );
        expect(code).toMatchSnapshot();
      });
    });
  }
});
