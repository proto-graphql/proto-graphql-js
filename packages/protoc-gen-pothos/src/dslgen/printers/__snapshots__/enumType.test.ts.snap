// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createEnumTypeCode > protobuf-es > 'generates code for a simple enum' 1`] = `
"import { EnumRef } from "@pothos/core";
import { MyEnum } from "@testapis/protobuf-es/testapis/enums/enums";
import { builder } from "../../builder";


export const MyEnum$Ref: EnumRef<MyEnum, MyEnum> =
  builder.enumType("MyEnum", {
    values: {
      FOO: {"value":1,"extensions":{"protobufEnumValue":{"name":"MY_ENUM_FOO"}}},
      BAR: {"description":"This is Bar.","value":2,"extensions":{"protobufEnumValue":{"name":"MY_ENUM_BAR"}}},
      BAZ: {"value":3,"extensions":{"protobufEnumValue":{"name":"MY_ENUM_BAZ"}}},
    } as const,
    extensions: {"protobufEnum":{"name":"MyEnum","fullName":"testapi.enums.MyEnum","package":"testapi.enums"}},
  });
"
`;

exports[`createEnumTypeCode > protobuf-es > 'generates code for an enum without un…' 1`] = `
"import { EnumRef } from "@pothos/core";
import { MyEnumWithoutUnspecified } from "@testapis/protobuf-es/testapis/enums/enums";
import { builder } from "../../builder";


export const MyEnumWithoutUnspecified$Ref: EnumRef<MyEnumWithoutUnspecified, MyEnumWithoutUnspecified> =
  builder.enumType("MyEnumWithoutUnspecified", {
    values: {
      FOO: {"value":0,"extensions":{"protobufEnumValue":{"name":"MY_ENUM_WITHOUT_UNSPECIFIED_FOO"}}},
      BAR: {"value":1,"extensions":{"protobufEnumValue":{"name":"MY_ENUM_WITHOUT_UNSPECIFIED_BAR"}}},
      BAZ: {"value":2,"extensions":{"protobufEnumValue":{"name":"MY_ENUM_WITHOUT_UNSPECIFIED_BAZ"}}},
    } as const,
    extensions: {"protobufEnum":{"name":"MyEnumWithoutUnspecified","fullName":"testapi.enums.MyEnumWithoutUnspecified","package":"testapi.enums"}},
  });
"
`;

exports[`createEnumTypeCode > protobuf-es > 'generates code for nested enum' 1`] = `
"import { EnumRef } from "@pothos/core";
import { ParentMessage_NestedEnum } from "@testapis/protobuf-es/testapis/nested/nested";
import { builder } from "../../builder";


export const ParentMessageNestedEnum$Ref: EnumRef<ParentMessage_NestedEnum, ParentMessage_NestedEnum> =
  builder.enumType("ParentMessageNestedEnum", {
    values: {
      FOO: {"value":1,"extensions":{"protobufEnumValue":{"name":"FOO"}}},
      BAR: {"value":2,"extensions":{"protobufEnumValue":{"name":"BAR"}}},
    } as const,
    extensions: {"protobufEnum":{"name":"NestedEnum","fullName":"testapis.nested.ParentMessage.NestedEnum","package":"testapis.nested"}},
  });
"
`;

exports[`createEnumTypeCode > ts-proto > 'generates code for a simple enum' 1`] = `
"import { EnumRef } from "@pothos/core";
import { MyEnum } from "@testapis/ts-proto/testapis/enums/enums";
import { builder } from "../../builder";


export const MyEnum$Ref: EnumRef<MyEnum, MyEnum> =
  builder.enumType("MyEnum", {
    values: {
      FOO: {"value":1,"extensions":{"protobufEnumValue":{"name":"MY_ENUM_FOO"}}},
      BAR: {"description":"This is Bar.","value":2,"extensions":{"protobufEnumValue":{"name":"MY_ENUM_BAR"}}},
      BAZ: {"value":3,"extensions":{"protobufEnumValue":{"name":"MY_ENUM_BAZ"}}},
    } as const,
    extensions: {"protobufEnum":{"name":"MyEnum","fullName":"testapi.enums.MyEnum","package":"testapi.enums"}},
  });
"
`;

exports[`createEnumTypeCode > ts-proto > 'generates code for an enum without un…' 1`] = `
"import { EnumRef } from "@pothos/core";
import { MyEnumWithoutUnspecified } from "@testapis/ts-proto/testapis/enums/enums";
import { builder } from "../../builder";


export const MyEnumWithoutUnspecified$Ref: EnumRef<MyEnumWithoutUnspecified, MyEnumWithoutUnspecified> =
  builder.enumType("MyEnumWithoutUnspecified", {
    values: {
      FOO: {"value":0,"extensions":{"protobufEnumValue":{"name":"MY_ENUM_WITHOUT_UNSPECIFIED_FOO"}}},
      BAR: {"value":1,"extensions":{"protobufEnumValue":{"name":"MY_ENUM_WITHOUT_UNSPECIFIED_BAR"}}},
      BAZ: {"value":2,"extensions":{"protobufEnumValue":{"name":"MY_ENUM_WITHOUT_UNSPECIFIED_BAZ"}}},
    } as const,
    extensions: {"protobufEnum":{"name":"MyEnumWithoutUnspecified","fullName":"testapi.enums.MyEnumWithoutUnspecified","package":"testapi.enums"}},
  });
"
`;

exports[`createEnumTypeCode > ts-proto > 'generates code for enum with extensio…' 1`] = `
"import { EnumRef } from "@pothos/core";
import { PrefixedEnum } from "@testapis/ts-proto/testapis/extensions/extensions";
import { builder } from "../../builder";


export const TestPrefixPrefixedEnum$Ref: EnumRef<PrefixedEnum, PrefixedEnum> =
  builder.enumType("TestPrefixPrefixedEnum", {
    values: {
      PREFIXED_FOO: {"value":1,"extensions":{"protobufEnumValue":{"name":"PREFIXED_FOO"}}},
      PREFIXED_BAR: {"value":2,"extensions":{"protobufEnumValue":{"name":"PREFIXED_BAR"}}},
    } as const,
    extensions: {"protobufEnum":{"name":"PrefixedEnum","fullName":"testapis.extensions.PrefixedEnum","package":"testapis.extensions"}},
  });
"
`;

exports[`createEnumTypeCode > ts-proto > 'generates code for nested enum' 1`] = `
"import { EnumRef } from "@pothos/core";
import { ParentMessage_NestedEnum } from "@testapis/ts-proto/testapis/nested/nested";
import { builder } from "../../builder";


export const ParentMessageNestedEnum$Ref: EnumRef<ParentMessage_NestedEnum, ParentMessage_NestedEnum> =
  builder.enumType("ParentMessageNestedEnum", {
    values: {
      FOO: {"value":1,"extensions":{"protobufEnumValue":{"name":"FOO"}}},
      BAR: {"value":2,"extensions":{"protobufEnumValue":{"name":"BAR"}}},
    } as const,
    extensions: {"protobufEnum":{"name":"NestedEnum","fullName":"testapis.nested.ParentMessage.NestedEnum","package":"testapis.nested"}},
  });
"
`;
